Match insignificant strides for sdpa inputs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124859
* #124751

Fix for https://github.com/pytorch/pytorch/issues/124289.

There was a tensor which had a single, expanded element. inductor generated the strides as all 0, while sdpa expects a dense last dimension `t.stride(-1) == 1`. While these are equivalent, we still hit an error in the kernel. We could make fixes in sdpa, but matching the insignificant strides in inductor also works and I am less aware of the downstream sdpa kernel details. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Match insignificant strides for sdpa inputs

[ghstack-poisoned]

eellison committed:
Update on "Match insignificant strides for sdpa inputs"


Fix for https://github.com/pytorch/pytorch/issues/124289.

There was a tensor which had a single, expanded element. inductor generated the stridse as all 0, while sdpa expects a dense last dimension `t.stride(-1) == 1`. While these are equivalent we were hitting an error. We could make fixes in sdpa, but matching the insignificant strides in inductor also works and I am less aware of the downstream sdpa kernel details. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented on test/inductor/test_fused_attention.py:
```diff
@@ -160,6 +161,81 @@ def dot_prod_attention(
                     check_train=False,
                 )
 
+    def _test_insignificant_strides(self):
```
nit: its not immediately clear what this is testing from the input

drisspg reviewed (approved):
stamp

eellison committed:
Update on "Match insignificant strides for sdpa inputs"


Fix for https://github.com/pytorch/pytorch/issues/124289.

There was a tensor which had a single, expanded element. inductor generated the stridse as all 0, while sdpa expects a dense last dimension `t.stride(-1) == 1`. While these are equivalent we were hitting an error. We could make fixes in sdpa, but matching the insignificant strides in inductor also works and I am less aware of the downstream sdpa kernel details. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Match insignificant strides for sdpa inputs"


Fix for https://github.com/pytorch/pytorch/issues/124289.

There was a tensor which had a single, expanded element. inductor generated the stridse as all 0, while sdpa expects a dense last dimension `t.stride(-1) == 1`. While these are equivalent we were hitting an error. We could make fixes in sdpa, but matching the insignificant strides in inductor also works and I am less aware of the downstream sdpa kernel details. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge --ignore-current

jithunnair-amd commented:
@eellison I believe this PR introduced a breakage for ROCm because mem_efficient_attention is not yet supported on ROCm, and this PR introduced an unchecked dependency on it via https://github.com/pytorch/pytorch/commit/68225072e864afcbc3cf0e4a078f117c9ab582b6#diff-56f796dea74c470ffe94a8517ddade6c99008c54f74b3f13bf1486b23d2a8181R227?

Should the code use `PLATFORM_SUPPORTS_MEM_EFF_ATTENTION` instead of `PLATFORM_SUPPORTS_FUSED_ATTENTION` here: https://github.com/pytorch/pytorch/commit/68225072e864afcbc3cf0e4a078f117c9ab582b6#diff-56f796dea74c470ffe94a8517ddade6c99008c54f74b3f13bf1486b23d2a8181R910?

eellison commented:
@jithunnair-amd yes you are correct

jithunnair-amd commented:
Alternately, if my suggested change would unnecessarily disable unit tests for ROCm that do *not* really depend on mem_efficient_attention, maybe a more targeted guard would be better.

eellison commented:
pr here: https://github.com/pytorch/pytorch/pull/124968
